### ISSUE TITLE
Proposal for alternative map-friendly block grammar

### ISSUE THREAD

**Issue by alexharv074** [hashicorp/terraform#22889]
### Summary

I wish to propose an alternative map-friendly grammar for nested blocks. It should be allowed to work alongside the existing grammar so as not to break backwards compatibility.

### Example

Instead of (legacy):

```tf
resource "aws_security_group" "web_traffic" {
  name        = "web_traffic"
  description = "Allow inbound traffic"

  ingress {
    from_port   = 80
    to_port     = 80
    protocol    = "-1"
    cidr_blocks = "10.0.0.0/8"
  }

  ingress {
    from_port   = 443
    to_port     = 443
    protocol    = "-1"
    cidr_blocks = "10.0.0.0/8"
  }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}
```

Also allow:

```tf
resource "aws_security_group" "web_traffic" {
  name        = "web_traffic"
  description = "Allow inbound traffic"

  ingress     = [   // PROPOSAL ONLY. DOES NOT ACTUALLY WORK !!
    {
      from_port   = 80
      to_port     = 80
      protocol    = "-1"
      cidr_blocks = "10.0.0.0/8"
    },
    {
      from_port   = 443
      to_port     = 443
      protocol    = "-1"
      cidr_blocks = "10.0.0.0/8"
    },
  ]

  egress     = [{
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }]
}
```

### Use-cases

Thanks to the otherwise awesome for_each feature that allows resources to be generated from data, it is now possible to generate resources from maps that have the same data structure as the generated resources.

The exception to this is resources that contain nested blocks, because a resource with a nested block cannot be represented in a map. This is because a map cannot contain duplicate keys, whereas duplicate block names are allowed in resources.

### Example usage

Define some data:

```tf
locals {
  aws_security_groups = {
    web_traffic = {
      description = "Allow inbound traffic"
      ingress     = [
        {
          from_port   = 80
          to_port     = 80
          protocol    = "-1"
          cidr_blocks = "10.0.0.0/8"
        },
        {
          from_port   = 80
          to_port     = 80
          protocol    = "-1"
          cidr_blocks = "10.0.0.0/8"
        },
      ]
      egress = {
        from_port   = 0
        to_port     = 0
        protocol    = "-1"
        cidr_blocks = ["0.0.0.0/0"]
      }
    }
  }
}
```

Code generate the resources:

```tf
resource "aws_security_group" "aws_security_groups" {
  for_each    = local.aws_security_groups
  name        = each.key
  description = each.value.description
  ingress     = each.value.ingress
  egress      = each.value.egress
}
```

### Also consider implementing splat grammar

Based on what is also possible in Puppet, this could be further improved if splat grammar was implemented, i.e. allow this instead:

```tf
resource "aws_security_group" "aws_security_groups" {
  for_each = local.aws_security_groups
  name = each.key
  *    = each.value 
}
```

This is a Terraform-isation of what we would write if this was the Puppet DSL, which would be:

```puppet
$aws_security_groups.each |$key,$value| {
  aws_security_group { $key:
    * => $value
  }
}
```

### References

- my blog post [here](https://alexharv074.github.io/2019/09/17/adventures-in-the-terraform-dsl-part-vii-resource-for_each-in-terraform-0.12.6.html).

**Comment by teamterraform** [comment]
Hi @alexharv074,

The mechanism for dynamic nested blocks in the Terraform language is [`dynamic` blocks](https://www.terraform.io/docs/configuration/expressions.html#dynamic-blocks), which are built from the nested block syntax in order to allow mixing both static and dynamic blocks at the same level without changing the syntax of the static blocks:

```hcl
resource "aws_security_group" "web_traffic" {
  name        = "web_traffic"
  description = "Allow inbound traffic"

  ingress {
    from_port   = 80
    to_port     = 80
    protocol    = "-1"
    cidr_blocks = "10.0.0.0/8"
  }

  dynamic "ingress" {
    for_each = var.additional_ingress_rules
    content {
      from_port   = ingress.value.from_port
      to_port     = ingress.value.to_port
      protocol    = ingress.value.protocol
      cidr_blocks = ingress.value.cidr_blocks
    }
  }
}
```

You can combine the resource-level `for_each` feature with `dynamic` blocks in order to construct resource instances fully dynamically if desired, though we do tend to caution against over-generalization because it can lead to reduced readability in the resulting configuration.

The omission of a feature to dynamically populate arguments inside a block from a map was an intentional language design decision that aims to prioritize the future reader of the configuration: requiring all of the arguments to be enumerated explicitly inside blocks means that the reader can easily see which arguments are set and which are not just from reading the block, rather than having to track down all of the expressions that contributed to the map/object. The explicit enumeration of the arguments also serves as a statement of intent from the author of which map keys they _intended_ to be present in the map, allowing Terraform to check that assumption and return an error if that assumption is violated.

This does mean that it is not possible to create fully-generic wrapper modules around arbitrary resource types. That was an intentional design tradeoff and not something we intend to support; modules should raise the level of abstraction by defining a higher-level concept, not simply pass through the full details of some other concept already defined in a provider. In your example with an `aws_security_group` resource whose number of instances and content is entirely dynamic, we would suggest not writing such a construct at all and instead just writing out literal `aws_security_group` resource blocks that directly describe the intended infrastructure; this resource block is not creating any higher-level abstraction, and is therefore redundant. There's more information on this design goal in the documentation section [When to write a module](https://www.terraform.io/docs/modules/index.html#when-to-write-a-module).

The Puppet DSL _does_ have some different design goals and is, speaking at a very high level, a more dynamic language than the Terraform language. Terraform leans towards explicit, static configuration as its default posture in order to make it as easy as possible to walk backwards from an object we see in production to understand the plan that last updated it and the configuration that declared it. There are some carefully-considered dynamic features in the language to allow creating lightweight abstractions, but overall the language design does not aspire to dynamic behaviors.

The existing `dynamic` blocks feature should allow you to meet the use-case you stated here, of constructing nested blocks dynamically from a list, set, or map of objects. Hopefully that helps you achieve your direct goal here, even though the syntax isn't exactly what you proposed.

**Comment by alexharv074** [comment]
@teamterraform , abstractions are usually needed when the alternative is copy/paste re-use. Static configurations are great until you find you have to copy/paste them into 10 different AWS accounts. I came upon this requirement in the context of a long list of default config rules in the AWS Config service. How do I define a list of default config rules while allowing the user to optionally disable them if they wish to provide a list of rules of their own. Puppet offers many solutions to this problem. I could put the rules in a class and then optionally include or not include that class. But Terraform has no concept of a class. I could require all accounts to simply define their own rules. But that results in duplication of a long list of rules as mentioned. I could provide a boolean variable use_default_rules but that would requirement to repeat the fairly unreadable line, "count = var.use_default_rules ? 1 : 0" about 50 times. I could use either of those solutions Puppet offers - a class, or a map of default resources - but Terraform offers neither. I can't see any good solution to this at the moment.

**Comment by apparentlymart** [comment]
Hello again! Sorry for the long silence.

Terraform's provider protocol version 6 introduced the possibility for provider developers to declare arguments which expect what we call "structural types", part of which is the ability to declare something which has schema capabilities similar to a nested block but syntax based on the object constructor syntax.

Terraform Core cannot force existing providers to use that new syntax unilaterally, because it introduces some new situations providers must deal with: the entire collection could be unknown or null. Those situations were impossible for nested blocks and so existing provider logic cannot typically handle it.

Providers must therefore opt in to using this new syntax, which is a gradual process happening outside the scope of this repository. Provider developers that wish to make use of this syntax must first build the relevant resource types using [Terraform Plugin Framework](https://www.terraform.io/plugin/framework), which is a new SDK not saddled by the limitations of the Terraform v0.11-and-earlier type system, and can then use `schema.SingleNestedAttributes`, `schema.ListNestedAttributes`, or `schema.MapNestedAttributes` to declare the structurally-typed arguments.

Because we use issues in this repository only to track work in this repository, and the protocol and Terraform Core parts of this are already implemented (since Terraform v1.1 or thereabouts), I'm going to close this issue. The adoption of this new capability is outside of the control of the Terraform Core team. In particular, adoption of the new syntax for the `ingress` and `egress` blocks mentioned in this issue is a decision to be made by the AWS provider team.

Thanks!
